R/rvar-.R

```diff
@@ -0,0 +1,1066 @@
+#' Random variables of arbitrary dimension
+#'
+#' Random variables backed by arrays of arbitrary dimension
+#'
+#' @name rvar
+#'
+#' @param x A vector or array where the first dimension represents draws from
+#' a distribution. The resulting [`rvar`] will have dimension `dim(x)[-1]`; that is,
+#' everything except the first dimension is used for the shape of the variable, and the
+#' first dimension is used to index draws from the distribution.
+#' @template args-rvar-dim
+#' @template args-rvar-dimnames
+#' @template args-format-nchains
+#'
+#' @details
+#'
+#' The `"rvar"` class internally represents random variables as arrays of arbitrary
+#' dimension, where the first dimension is used to index draws from the distribution.
+#
+#' Most mathmetical operators and functions are supported, including efficient matrix
+#' multiplication and vector and array-style indexing. The intent is that an `rvar`
+#' works as closely as possible to how a base vector/matrix/array does, with a few
+#' differences:
+#'
+#' - The default behavior when subsetting is not to drop extra dimensions (i.e.
+#'   the default `drop` argument for `[` is `FALSE`, not `TRUE`).
+#' - Rather than base R-style recycling, `rvar`s use a limited form of broadcasting:
+#'   if an operation is being performed on two vectors with different size of the same
+#'   dimension, the smaller vector will be recycled up to the size of the larger one
+#'   along that dimension so long as it has size 1.
+#'
+#' For functions that expect base numeric arrays and for which `rvar`s cannot be
+#' used directly as arguments, you can use [rfun()] or [rdo()] to translate your
+#' code into code that executes across draws from one or more random variables
+#' and returns a random variable as output. Typically [rdo()] offers the most
+#' straightforward translation.
+#'
+#' As [rfun()] and [rdo()] incur some performance cost, you can also operate directly
+#' on the underlying array using the [draws_of()] function.
+#'
+#' @seealso [as_rvar()] to convert objects to `rvar`s.
+#'
+#' @return An object of class `"rvar"` representing a random variable.
+#'
+#' @export
+rvar <- function(x = double(), dim = NULL, dimnames = NULL, .nchains = 1L) {
+  x <- new_rvar(x, .nchains = .nchains)
+
+  if (!is.null(dim)) {
+    dim(x) <- dim
+  }
+  if (!is.null(dimnames)) {
+    dimnames(x) <- dimnames
+  }
+
+  x
+}
+
+#' @importFrom vctrs new_vctr
+new_rvar <- function(x = double(), .nchains = 1L) {
+  # TODO: decide on supported types and cast to them in here
```

Comment:
No, it shouldn't cause any problems to defer it. 

At some point it would probably be good to set up proper casting to/from the {distributional} data types, but I don't think it's necessary for the first version. I'll open a separate issue and remove the TODO from here.